tonic/src/transport/service/timeout.rs

```diff
@@ -0,0 +1,251 @@
+        }
+
+        Poll::Pending
+    }
+}
+
+#[pin_project(project = OptionPinProj)]
+enum OptionPin<T> {
+    Some(#[pin] T),
+    None,
+}
+
+const GRPC_TIMEOUT_HEADER: &str = "grpc-timeout";
+
+const SECONDS_IN_HOUR: u64 = 60 * 60;
+const SECONDS_IN_MINUTE: u64 = 60;
+
+/// Tries to parse the `grpc-timeout` header if it is present. If we fail to parse, returns
+/// the value we attempted to parse.
+///
+/// Follows the [gRPC over HTTP2 spec](https://github.com/grpc/grpc/blob/master/doc/PROTOCOL-HTTP2.md).
+fn try_parse_grpc_timeout(
+    headers: &HeaderMap<HeaderValue>,
+) -> Result<Option<Duration>, &HeaderValue> {
+    match headers.get(GRPC_TIMEOUT_HEADER) {
+        Some(val) => {
+            let (timeout_value, timeout_unit) = val
+                .to_str()
+                .map_err(|_| val)
+                .and_then(|s| if s.is_empty() { Err(val) } else { Ok(s) })?
+                // `HeaderValue::to_str` only returns `Ok` if the header contains ASCII so this
+                // `split_at` will never panic from trying to split in the middle of a character.
+                // See https://docs.rs/http/0.2.4/http/header/struct.HeaderValue.html#method.to_str
+                //
+                // `len - 1` also wont panic since we just checked `s.is_empty`.
+                .split_at(val.len() - 1);
+
+            // gRPC spec specifies `TimeoutValue` will be at most 8 digits
+            // Caping this at 8 digits also prevents integer overflow from ever occurring
+            if timeout_value.len() > 8 {
+                return Err(val);
+            }
+
+            let timeout_value: u64 = timeout_value.parse().map_err(|_| val)?;
+
+            let duration = match timeout_unit {
+                // Hours
+                "H" => Duration::from_secs(timeout_value * SECONDS_IN_HOUR),
+                // Minutes
+                "M" => Duration::from_secs(timeout_value * SECONDS_IN_MINUTE),
+                // Seconds
+                "S" => Duration::from_secs(timeout_value),
+                // Milliseconds
+                "m" => Duration::from_millis(timeout_value),
+                // Microseconds
+                "u" => Duration::from_micros(timeout_value),
+                // Nanoseconds
+                "n" => Duration::from_nanos(timeout_value),
+                _ => return Err(val),
+            };
+
+            Ok(Some(duration))
+        }
+        None => Ok(None),
+    }
+}
+
+// Note: The wrapped Duration should only be used for logging purposes. It is **not** the
+// actual duration that elapsed, resulting in a timeout, instead it is a close approximation
+#[derive(Debug)]
+struct TimeoutExpired;
+
+impl fmt::Display for TimeoutExpired {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        write!(f, "Timeout expired")
+    }
+}
+
+// std::error::Error only requires a type to impl Debug and Display
+impl std::error::Error for TimeoutExpired {}
+
+#[cfg(test)]
+mod tests {
```

Comment:
Might be interesting to add a quicktest fuzz test here?

Reply:
@davidpdrsn if you don't have time for this, I've wanted to mess around with `quicktest` but haven't had a chance yet. Would be more than happy to add a follow up diff adding a fuzz test here 🙂

Reply:
I've used quickcheck before so I'll look at it now 😊

Reply:
Implemented in 3452d3077653f.